### PR TITLE
multiple RP support -2: reset controller for multiple RPs

### PR DESCRIPTION
**Pull request by yb01**
**What type of PR is this?**
> /kind feature

**What this PR does / why we need it**:
reset kube controller manager on TPs with kubeconfig files from RP clusters

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Testing:
2 TP 2 RP scaleout:

```
export RUN_PREFIX=yb01-multiple-rp

export MASTER_SIZE=n1-highmem-32 NUM_NODES=2 KUBEMARK_NUM_NODES=100

export MASTER_DISK_SIZE=200GB MASTER_ROOT_DISK_SIZE=200GB KUBE_GCE_ZONE=us-central1-a NODE_SIZE=n1-highmem-16 NODE_DISK_SIZE=200GB KUBE_GCE_ENABLE_IP_ALIASES=true KUBE_GCE_PRIVATE_CLUSTER=true CREATE_CUSTOM_NETWORK=true ETCD_QUOTA_BACKEND_BYTES=8589934592 TEST_CLUSTER_LOG_LEVEL=--v=2 ENABLE_KCM_LEADER_ELECT=false ENABLE_SCHEDULER_LEADER_ELECT=false SHARE_PARTITIONSERVER=false APISERVERS_EXTRA_NUM=0 WORKLOADCONTROLLER_EXTRA_NUM=0 ETCD_EXTRA_NUM=0 LOGROTATE_FILES_MAX_COUNT=50 LOGROTATE_MAX_SIZE=200M KUBE_GCE_INSTANCE_PREFIX=${RUN_PREFIX} KUBE_GCE_NETWORK=${RUN_PREFIX}  CREATE_TEST_TENANTS=true
export TEST_CLUSTER_LOG_LEVEL=--v=4
export KUBE_ENABLE_APISERVER_INSECURE_PORT=true
export KUBE_ENABLE_PROMETHEUS_DEBUG=true
export SCALEOUT_CLUSTER=true
export SCALEOUT_TP_COUNT=2
export SCALEOUT_RP_COUNT=2

DBG: RP number: 2
DBG: PR kubeconfig: /home/cloudshare/go/src/k8s.io/kubernetes/test/kubemark/resources/kubeconfig.kubemark.rp-2
Waiting for all hollow-nodes to become Running.................. Done!

Master IP: 23.251.147.249

Listing kubeamrk cluster details:
Getting total nodes number:
102

Getting total hollow-nodes number:
100


Getting total hollow-nodes number for RP-1
100


Getting total hollow-nodes number for RP-2
100

Getting endpoints status:
NAMESPACE     NAME                   ENDPOINTS       AGE   SERVICEGROUPID
default       kubernetes             10.40.0.7:443   13m   0
kube-system   kubernetes-dashboard   <none>          13m   <none>

Getting workload controller co status:
No resources found.

Getting apiserver data partition status:
No resources found.

Getting ETCD data partition status:
No resources found.


```

Test with 1TP and 1RP:
```
I0322 23:09:41.989822   12069 simple_test_executor.go:385] Resources cleanup time: 15.015967957s
I0322 23:09:41.989885   12069 clusterloader.go:218] --------------------------------------------------------------------------------
I0322 23:09:41.989891   12069 clusterloader.go:219] Test Finished
I0322 23:09:41.989895   12069 clusterloader.go:220]   Test: testing/density/config.yaml
I0322 23:09:41.989900   12069 clusterloader.go:221]   Status: Success
I0322 23:09:41.989905   12069 clusterloader.go:225] --------------------------------------------------------------------------------

JUnit report was created: /home/cloudshare/logs/test/junit.xml

Pod startup time:
{
      "data": {
        "Perc50": 1799.404097,
        "Perc90": 2448.325778,
        "Perc99": 2905.017265
      },
      "unit": "ms",
      "labels": {
        "Metric": "pod_startup"
      }
    },
```

Test with 2 TP and 2 RP:
```
I0323 01:54:01.486842    2160 simple_test_executor.go:385] Resources cleanup time: 15.015980038s
I0323 01:54:01.486907    2160 clusterloader.go:218] --------------------------------------------------------------------------------
I0323 01:54:01.486912    2160 clusterloader.go:219] Test Finished
I0323 01:54:01.486917    2160 clusterloader.go:220]   Test: testing/density/config.yaml
I0323 01:54:01.486921    2160 clusterloader.go:221]   Status: Success
I0323 01:54:01.486926    2160 clusterloader.go:225] --------------------------------------------------------------------------------

JUnit report was created: /home/cloudshare/logs/test/junit.xml
```

Test scale up:
```
unset SCALEOUT_CLUSTER
unset SCALEOUT_TP_COUNT
unset SCALEOUT_RP_COUNT

### kube-up.sh
### start-kubemark.sh
### perf test run:
cloudshare@ybtest-11:~/go/src/k8s.io/kubernetes$ 
[1]+  Done                    ./perf-tests/clusterloader2/run-e2e.sh --nodes=100 --provider=kubemark --kubeconfig=/home/cloudshare/go/src/k8s.io/kubernetes/test/kubemark/resources/kubeconfig.kubemark --report-dir=/home/cloudshare/logs/test --testconfig=testing/density/config.yaml > testscaleup.log 2>&1
cloudshare@ybtest-11:~/go/src/k8s.io/kubernetes$ 
I0323 05:26:55.917391     320 clusterloader.go:218] --------------------------------------------------------------------------------
I0323 05:26:55.917395     320 clusterloader.go:219] Test Finished
I0323 05:26:55.917400     320 clusterloader.go:220]   Test: testing/density/config.yaml
I0323 05:26:55.917404     320 clusterloader.go:221]   Status: Success
I0323 05:26:55.917408     320 clusterloader.go:225] --------------------------------------------------------------------------------

JUnit report was created: /home/cloudshare/logs/test/junit.xml
```

### CODE REVIEW

File: cmd/kube-controller-manager/app/options/options.go [hunk 1]
```diff
@@ -423,16 +423,19 @@ func (s KubeControllerManagerOptions) Config(allControllers []string, disabledBy
...
 		resourceProviderKubeConfigFiles, existed := genutils.ParseKubeConfigFiles(s.ResourceProviderKubeConfig)
 		if !existed {
-			return nil, fmt.Errorf("--resource-providers points to non existed file(s)")
-		}
-		resourceProviderClients = make([]clientset.Interface, len(resourceProviderKubeConfigFiles))
-		for i, kubeConfigFile := range resourceProviderKubeConfigFiles {
-			resourceProviderClient, err := clientutil.CreateClientFromKubeconfigFile(kubeConfigFile)
-			if err != nil {
-				return nil, fmt.Errorf("failed to create resource provider rest client from kubeconfig [%s], error [%v]", kubeConfigFile, err)
+			klog.Warningf("--resource-providers points to non existed file(s), default to local cluster kubeconfig file")
```
Comment (Sindica) [inline cmd/kube-controller-manager/app/options/options.go:426]: Can we add a TODO here? Thanks.

File: cmd/kube-scheduler/app/options/options.go [hunk 2]
```diff
@@ -264,23 +264,28 @@ func (o *Options) Config() (*schedulerappconfig.Config, error) {
...
-		c.NodeInformers = make(map[string]coreinformers.NodeInformer, len(kubeConfigFiles))
-		for i, kubeConfigFile := range kubeConfigFiles {
-			rpId := "rp" + strconv.Itoa(i)
-			c.ResourceProviderClients[rpId], err = clientutil.CreateClientFromKubeconfigFile(kubeConfigFile)
-			if err != nil {
-				klog.Error("failed to create resource provider rest client, error: %v", err)
-				return nil, err
-			}
-
+			klog.Warningf("ResourceProvider kubeConfig is not valid, default to local cluster kubeconfig file")
```
Comment (Sindica) [inline cmd/kube-scheduler/app/options/options.go:267]: Same here

File: cluster/gce/gci/configure-helper-common.sh [hunk 3]
```diff
@@ -2341,6 +2341,16 @@ function start-kube-controller-manager {
    params+=" --kubeconfig=/etc/srv/kubernetes/kube-controller-manager/kubeconfig"
   fi
 
+   # the resource provider kubeconfig
+  if [[ "${KUBERNETES_TENANT_PARTITION:-false}" == "true" ]]; then
+    local rp_kubeconfigs="/etc/srv/kubernetes/kube-controller-manager/rp-kubeconfig-1"
```
[the remark this answers is not in the document]
Comment (yb01) [reply cluster/gce/gci/configure-helper-common.sh:2346]: added TODO and comments on design limitations in start-kubemark.sh where the files were copied.

File: cmd/kube-controller-manager/app/options/options.go (hunk 1 above)
Comment (yb01) [reply cmd/kube-controller-manager/app/options/options.go:428]: added TODO for design limitations.

File: cmd/kube-scheduler/app/options/options.go (hunk 2 above)
Comment (yb01) [reply cmd/kube-scheduler/app/options/options.go:269]: added

**Review by Sindica — APPROVED** [review]
/lgtm

**Pull request merged**